recommender/engine/rl_agent/models/actor_model.py

```diff
@@ -1,23 +1,56 @@
+# pylint: disable=missing-module-docstring, invalid-name, too-many-arguments, no-member
+
+from typing import Tuple
+
 import torch
+from torch import nn
+import torch.nn.functional as F
+
+from recommender.engine.rl_agent.models.history_embedder import HistoryEmbedder
+
 
+class ActorModel(nn.Module):
+    """Actor neural network representing a deterministic policy used by RLAgent"""
 
-class ActorModel(torch.nn.Module):
-    def __init__(self, K: int, SE: int, history_embedder: torch.nn.Module):
+    def __init__(
+        self,
+        K: int,
+        SE: int,
+        UE: int,
+        FE: int,
+        SPE: int,
+        history_embedder: HistoryEmbedder,
+        layer_sizes: Tuple[int] = (256, 512, 256),
+    ):
+        """
+        Args:
+            K: number of services to recommend
+            SE: service embedding dimension
+            UE: user embedding dimension
+            FE: filter embedding dimension
+            SPE search phrase embedding dimension
```

Comment:
Search phrase is embedded as tensor of shape [X, SPE], where X is the number of words in the search phrase string and SPE is search phrase embedding dimension (of the word). So seachphrase tensor is two dimensional - not one dimensional.

Now we can have at least two possible solutions:

1. Build some kind of RNN into Actor - to reduce search phrase's tensor dimensions to SPE
2. Build SearchPhraseEmbedderModel (that is pytorch module) - that get tensor of shape [X, SPE] from SearchPhraseEmbedder (that is not pytorch module and just transform string to tensor) and reduce its dimensions using some RNN.

Solution 1 is wrong because the state (that contains embedded search phrase tensor of shape [X, SPE]) will be also provided to CriticModel, so it will have to have its own separate RNN for reducing dimensions. It is exactly the cause of separating out HistoryEmbedder and sharing its weights between ActorModel and CriticModel - but for history tensor. Now we should do the same thing for searchphrase tensor and separate out SearchPhraseEmbedderModel that weights should be shared between ActorModel and CriticModel - so soultion 2 is the good one.

To sum it up:
We should have:

- SearchPhraseEmbedder that is responsible for embedding search phrase string into search phrase tensor of shape [X, SPE]
- SearchPhraseEmbedderModel that is reponsible for reducing dimensions from [X, SPE] to [SPE] and its weights will be shared between ActorModel and Critic Model.

Reply:
I will create separate issue for SearchPhraseEmbedderModel

Reply:
I agree however we should work on better names - if HistoryEmbedder is a module, then the SearchPhraseEmbedder should also be a Module. I suggest we rename SearchPhraseEmbedder to SearchPhraseExtractor or Preprocessor